Validation batch update

Alexey-ebi opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Enable to have batch update for validation issues
## Description
<!--- Describe your changes in detail -->

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Put closes #XXXX in your comment to auto-close the issue that your -->
<!--- PR fixes (if such).Please link to the issue here: -->
partially address #16

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change which improve code itself)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Alexey-ebi committed:
Adding ids to messages for batch update of validation erros

Alexey-ebi committed:
Creating view for batch validation fix

Alexey-ebi committed:
Adding error message for fix error view

Alexey-ebi committed:
Providing forms for batch update

Alexey-ebi committed:
Fixing design

Alexey-ebi committed:
Adding celery task to run batch update

Alexey-ebi committed:
Fixing problem with hard-coded variables

Alexey-ebi committed:
Merging master branch

Alexey-ebi committed:
Fixing tests

bunop commented on django-data/image/animals/tasks.py:
```diff
@@ -0,0 +1,84 @@
+import asyncio
+
+from common.constants import ERROR, LOADED, STATUSES
+from common.helpers import send_message_to_websocket
```
`asyncio` and `send_message_to_websocket` import can be removed if using `submission.helpers.send_message`

Alexey-ebi replied:
Done

bunop commented on django-data/image/animals/tasks.py:
```diff
@@ -0,0 +1,84 @@
+                },
+                args[0]
+            )
+        )
```
```python
asyncio.get_event_loop().run_until_complete(
    send_message_to_websocket(
        {
            'message': STATUSES.get_value_display(ERROR),
            'notification_message': submission_obj.message
        },
        args[0]
    )
)
```
can be replaced with
```python
send_message(submission_obj)
```

Alexey-ebi replied:
Done

bunop commented on django-data/image/animals/tasks.py:
```diff
@@ -0,0 +1,84 @@
+        # TODO: submit mail to admin
+
+    def run(self, submission_id, animal_ids):
+        """a function to upload data into UID"""
```
Please update docstring

Alexey-ebi replied:
Done

bunop commented on django-data/image/animals/tasks.py:
```diff
@@ -0,0 +1,84 @@
+
+        # Update submission
+        submission_obj = Submission.objects.get(pk=submission_id)
+        submission_obj.status = LOADED
```
should be `NEED_REVISION`

Alexey-ebi replied:
I don't think this should be NEED_REVISION, as user already did revision and updated fields, so LOADED seems to be right status. LOADED status together with "Data updated, try to rerun validation" should be clear for user that it's required to rerun validation

bunop replied:
accordingly to [common.statuses](https://image-injecttool.readthedocs.io/en/latest/modules/common.html#common-statuses) `LOADED` is the status after data load, before doing any validation, while `NEED_REVISION` is the status which requires a validation. Status could return in `NEED_REVISION` for issues in biosample submission, data changes or a failure in validation or for any other type of issues with data iteself

Alexey-ebi replied:
ok, done

bunop commented on django-data/image/common/constants.py:
```diff
@@ -173,3 +173,18 @@ class SAMPLE_STORAGE_PROCESSING(EnumMixin, Enum):
 
 KNOWN_STATUSES = ['Error', 'Issues', 'Known', 'Pass', 'Warning']
+
+# TODO add all possible messages
```
If I had understood your point, you want to try to model a specific request for each validation message, and then launching a Task in which you will update a pre-determined set of column. I will suggest to model the offending field instead. Consider that Jun has modified validation code to have back the offending column in the `ValidationResult` method. Since I defined the conversion between UID fields into biosample fields in `image_app.models`, we could refactor this modulel to keep track of relations between biosample fields and UID fields. In such way, you don't need to hard code all the possible validation messages (that could change!) and you can derive the field to update in `BatchUpdateAnimals` and `BatchUpdateSamples` tasks using the biosample2uid key conversion table

bunop replied:
Consider that there are columns that InjectTools doesn't want modify: the relations beetween animals, sample and animal names and the relation between sample and animal id don't need to be changed in InjectTool or otherwise I will not able to determine what animal are into UID during reload, or People can do relationship between objects that don't own

Alexey-ebi replied:
In addition to offending field in current implementation I show helping fields (help user to decide which value to provide for offending field). So I agree that It would be great to store offending and helping fields in model. Also I could add new method (uid2biosamples for example) for Animal and Sample models that will convert human readable name to model name, and opposite. What do you think?

Alexey-ebi replied:
Done

Alexey-ebi committed:
Adding celery task for samples

bunop commented on django-data/image/animals/tasks.py:
```diff
@@ -0,0 +1,84 @@
+
+        for animal_id, value in animal_ids.items():
+            animal = Animal.objects.get(pk=animal_id)
+            animal.birth_location = value
```
Does this mean a specific task for every field we need to update? The attribute to update should be determined somewhere, for instance when instantiating a task

bunop replied:
Will be great to check also if such value is changed or not from the previous. If changed, update. To do here, not in views (we are in a task you could do any tedious stuff here)

Alexey-ebi replied:
Done

bunop commented on django-data/image/common/constants.py:
```diff
@@ -173,3 +173,44 @@ class SAMPLE_STORAGE_PROCESSING(EnumMixin, Enum):
+    'parents_sex_check': ''
+}
+
+VALIDATION_MESSAGES_ATTRIBUTES = {
```
In my opinion, It will be tedious defining every attribute combination to display. What about preset of fields (for instance, the 4 location columns for each location field to fix) or displaying more fields in the same line?

Alexey-ebi replied:
Done

bunop commented on django-data/image/submissions/templatetags/submissions_filters.py:
```diff
@@ -10,3 +12,20 @@ def get_dict_from_str(value):
+
+
+@register.filter
+def convert_to_human_readable(value):
```
This field name you are trying to render, is the verbose name of the field. Should we use a template tag for render field name (even in sample/animal detail page) as described [here](https://stackoverflow.com/a/14498938/4385116)?

Alexey-ebi replied:
Done

bunop commented on django-data/image/samples/tasks.py:
```diff
@@ -0,0 +1,84 @@
+        submission_obj.message = ("Error in batch update for animals: %s" % (str(exc)))
+        submission_obj.save()
+
+        asyncio.get_event_loop().run_until_complete(
```
Same considerations as Animal

Alexey-ebi replied:
Done

bunop commented on django-data/image/samples/tasks.py:
```diff
@@ -0,0 +1,84 @@
+
+        # Update submission
+        submission_obj = Submission.objects.get(pk=submission_id)
+        submission_obj.status = LOADED
```
Should be `NEED_REVISION`

Alexey-ebi replied:
Done

bunop commented on django-data/image/samples/tasks.py:
```diff
@@ -0,0 +1,84 @@
+logger = get_task_logger(__name__)
+
+
+class BatchUpdateSamples(MyTask):
```
Scrutinizer says that this code seems duplicated (and I think that is similar to `Animal`). Can we derive a common mixin for those tasks?

Alexey-ebi replied:
Done

bunop commented on django-data/image/samples/tasks.py:
```diff
@@ -0,0 +1,84 @@
+
+        for sample_id, value in sample_ids.items():
+            sample = Sample.objects.get(pk=sample_id)
+            sample.collection_place = value
```
Same considerations as Animals

Alexey-ebi replied:
Done

bunop commented on django-data/image/submissions/templates/submissions/submission_validation_summary.html:
```diff
@@ -19,7 +19,7 @@ <h1>Validation summary for {{ validation_summary.type }}s</h1>
       <span class="badge badge-warning badge-pill align-right">{{ message.count }}</span>
         {{ message.message }}
-        <a href="#"><span class="badge badge-success badge-pill align-right">Fix this</span></a>
+        <a href="{% url 'submissions:validation_summary_fix_errors' pk=submission.id type=validation_summary.type msg=message.message %}"><span class="badge badge-success badge-pill align-right">Fix this</span></a>
```
It is safe pass a validation message using a `GET` method? I suggest to use the field name instead, as commented in `common.constants` module

Alexey-ebi replied:
I've changed to pass ValidationSummary id and index of error in messages field

bunop commented on django-data/image/submissions/views.py:
```diff
@@ -351,3 +399,41 @@ def delete(self, request, *args, **kwargs):
+    # a valid submission start a task
+    res = my_task.delay(pk, keys_to_fix)
+    logger.info(
+        "Start crbanim importing process with task %s" % res.task_id)
```
Please, fix this message

Alexey-ebi replied:
Done

bunop commented on django-data/image/submissions/templates/submissions/submission_validation_summary_fix_errors.html:
```diff
@@ -0,0 +1,46 @@
+                    <li class="list-group-item list-group-item-warning d-flex justify-content-between align-items-center">
+                        <div class="from-group form-inline">
+                            <label for="{{ object.id|create_unique_id:'to_edit' }}">{{ show|convert_to_human_readable }}:</label>
+                            <input type="text" class="form-control" id="{{ object.id|create_unique_id:'to_edit' }}" value="{{ object|getattribute:show }}" name="{{ object.id|create_unique_id:'to_edit' }}">
```
`value="{{ object|getattribute:show }}"` will replaced by the string "None" for `birth_location` for example and recorded as a string in database. We need to deal the missing case or the fact that I would provide an empty record

bunop replied:
I understand that this implementation will be generic for the data we need to implement. However, we are missing the django built-ins regarding data field types, as the integer, date, or string types. Columns defined as foreign keys, such as the ontology terms, or the enum columns like accuracies are rendered as select types by django. Moreover, in addition to this simple form validation rendered by django (like don't put a character in a numeric field) django can do a complex validation, such don't put a `birth_location` with a *MISSING* `accuracy` in the same time. I don't kwnow if will be a good idea to write all fields and then hiding stuff using Javascript. Take this comment as a consideration, don' try to solve it forn now

Alexey-ebi replied:
Done

bunop commented on django-data/image/submissions/templatetags/submissions_filters.py:
```diff
@@ -10,3 +12,20 @@ def get_dict_from_str(value):
+@register.filter
+def getattribute(value, arg):
+    if arg == 'birth_location_accuracy' or arg == 'collection_place_accuracy':
+        return ACCURACIES.get_value_display(getattr(value, arg))
```
You don't need to access this field from common.constant.accuracies but accessing `get_birth_location_accuracy_display` and `get_collection_place_accuracy_display` as described for Enum columns

Alexey-ebi replied:
Done

bunop commented on django-data/image/validation/tasks.py:
```diff
@@ -56,8 +56,8 @@ def __init__(self, submission_obj, ruleset):
-        self.animals_messages = Counter()
-        self.samples_messages = Counter()
+        self.animals_messages = dict()
+        self.samples_messages = dict()
```
You could also use `collections.defaultdict`, this will change in: 
```python
self.animals_messages = defaultdict(list)
self.samples_messages = defaultdict(list)
```

Alexey-ebi replied:
Done

bunop commented on django-data/image/validation/tasks.py:
```diff
@@ -181,12 +181,14 @@ def mark_model(self, model, result, status):
                 # samples_messages is a counter object
-                self.samples_messages.update({message})
+                self.samples_messages.setdefault(message, [])
+                self.samples_messages[message].append(model.pk)
```
Using `collections.defaultdict` you could say 
```python
self.samples_messages[message].append(model.pk)
```
Whitout setting the default value (since is defined by `self.samples_messages = defaultdict(list)`)

Alexey-ebi replied:
Done

bunop commented on django-data/image/submissions/views.py:
```diff
@@ -351,3 +399,41 @@ def delete(self, request, *args, **kwargs):
+    submission.status = WAITING
+    submission.save()
+
+    # Update validation summary
```
Should we use a method like `validation.models.ValidationSummary.reset_all_count` to reset all the stuff to the default value? Should we change this method to does all the resets? Should we rename to `reset` or adding a new `reset`method?

Alexey-ebi replied:
I've added new reset method, that will reset all fields excluding all_count

bunop commented on django-data/image/submissions/views.py:
```diff
@@ -351,3 +399,41 @@ def delete(self, request, *args, **kwargs):
         return httpresponseredirect
+
+
+def fix_validation(request, pk, type):
```
Should this be a `CBV`? we could re-use the `OwnerMixin` to ensure data update for the right person

Alexey-ebi replied:
Done

bunop commented on django-data/image/submissions/templates/submissions/submission_validation_summary_fix_errors.html:
```diff
@@ -0,0 +1,46 @@
+    </div>
+    <form action="{% url 'submissions:fix_validation' submission.id type %}" method="post">
+        {% csrf_token %}
+        {% for object in  object_list %}
```
Could be an issue when having many objects

Alexey-ebi replied:
Here I will add pagination

bunop commented on django-data/image/common/constants.py:
```diff
@@ -173,3 +173,44 @@ class SAMPLE_STORAGE_PROCESSING(EnumMixin, Enum):
+    }
+}
+
+ATTRIBUTES_HUMAN_READABLE = {
```
Django does already this conversion, I suggest you a stakoverflow post to determine the `human_readable` name using a template tags. You wll see as a comment in `submission_filter.py` module. If you need, you could change what is displayed using the `verbose_name` attribute in UID models

Alexey-ebi replied:
Done

bunop commented on django-data/image/submissions/urls.py:
```diff
@@ -35,5 +35,11 @@
         views.SubmissionValidationSummaryView.as_view(),
-        name='validation_summary')
+        name='validation_summary'),
+    url(r'^(?P<pk>[-\w]+)/validation_summary/(?P<type>[\w]+)/(?P<msg>.*)/$',
```
Now we have more stuff for validation summary in submission views. I return to the idea to move all validation views in validation module. However this doesn't care at the moment, we can decide to move them later, if we have set up tests properly. For now keep things like this

Alexey-ebi replied:
ok

bunop commented on django-data/image/submissions/views.py:
```diff
@@ -155,16 +161,58 @@ class SubmissionValidationSummaryView(OwnerMixin, DetailView):
+        validation_summary = ValidationSummary.objects.get(
+            submission=self.submission, type=self.summary_type)
+        self.request_message = self.kwargs['msg']
+        for message in validation_summary.messages:
```
In ValidationSummary model, you have a field in which you store data about ids that have the same validation message, and I suggest you to track also the offending field to fix. You could also refactor validation models to add new related objects if this will simplifies your coding approach

Alexey-ebi replied:
Done

bunop commented on django-data/image/submissions/views.py:
```diff
@@ -155,16 +161,58 @@ class SubmissionValidationSummaryView(OwnerMixin, DetailView):
+        context["message"] = self.request_message
+        context["type"] = self.summary_type
+
+        if re.search(VALIDATION_MESSAGES['coordinate_check'],
```
See my consideration on `common.constants`

Alexey-ebi replied:
Done

bunop reviewed (changes requested):
Ty @Alexey-ebi for your contribution! Honestly, I expected a single form page in which all fields are rendered and the update is done on all selected objects. However I like the possibility to set a different value for each item but I understand that this mean addictional work, at least to pre-determine the combination of field/attributes to show/update and a small refactoring in order to re-use the code you have already wrote. I understood that the main use of bulk update will be setting the same attribute to the selected fields, however your approach is more reliable since it force the user to customize the attribute for each selected object. An open question is if a user will update one-by-one one an hundred of problematic fields and how this page rendering will be confortable to the user, but this is inclear to me at the moment. I'm only worried if an approach like this will be difficult to maintain and if it will require addictional work for example if we introduce the possibility to cherry-pick the objects to update or the possibility to bulk update objects from an object list uploaded by the user. If you think that the code maintanance will be minimal, I will accept your PR. I've added some comments on your code. Thank you again!

Alexey-ebi committed:
Refactoring code

Alexey-ebi committed:
Merging with master branch

Alexey-ebi committed:
Code refactoring

Alexey-ebi committed:
Starting to refactor code after code review

Alexey-ebi committed:
Refactoring code

Alexey-ebi committed:
Code refactoring

Alexey-ebi committed:
Adding offending collumn in model

Alexey-ebi committed:
Refactoring code

Alexey-ebi committed:
Code refactoring

Alexey-ebi committed:
Code refactoring and test fixing

Alexey-ebi committed:
Adding mixin to batch update tasks

Alexey-ebi committed:
Fixing bug with None value

Alexey-ebi committed:
Rewriting fix_validation in CBV

Alexey-ebi committed:
Adding reset method

Alexey-ebi committed:
Adding condition to show fix this button

Alexey-ebi committed:
Adding fix units validation handler

Alexey-ebi committed:
Adding message handler for availability errors

Alexey-ebi committed:
Adding handlers for values issues

Alexey-ebi committed:
Adding handlers for ontology validation errors

Alexey-ebi committed:
Code refactoring

Alexey-ebi committed:
Adding handlers for species

Alexey-ebi committed:
Adding offending_column to fixture

bunop commented on django-data/image/common/tasks.py:
```diff
@@ -13,10 +13,53 @@
+            elif item_type == 'animal':
+                item_object = Animal.objects.get(pk=id)
+            elif item_type == 'dictuberon':
+                item_object = DictUberon.objects.get(pk=id)
```
Are we sure that we want to change dictionary tables using batch updates? consider that there is one entry for each dictionary types, so you don't need to change a value in batches. Consider also that dictionary tables are shared accross each user and changing a value in this tables will affect all data, since the biosample conversion is done on the fly. This means that if you set a wrong ontology term, you  will invalidate my data, even if I'm submitting since I had valid data before your update. In my opinion, batch update should be valid only for Animal and Samples and for fields related to those models. I think that it will decrease your code complexity

bunop replied:
Maybe instead of changing the term value of those dictionary tables, we can set the relation between this table: for instance I want to change sex for an animal, for example. This relation is contained in animal, so it is safe to change this in animal. But changing the term value or ontology term should be done by a different process

Alexey-ebi replied:
Removed possibility to change ontology terms, just show message to user. This should be changed in another process using Zooma.

bunop commented on django-data/image/common/helpers.py:
```diff
@@ -152,3 +152,14 @@ def get_admin_emails():
     return [admin[1] for admin in ADMINS]
+
+
+def uid2biosample(value):
```
I mean a more complex class, since I need also to convert UID fields into Biosample fields, however is not important now. Ok for this function since it works, I think I will refactor this later if I return on UID database

Alexey-ebi replied:
ok

bunop commented on django-data/image/common/constants.py:
```diff
@@ -173,3 +173,34 @@ class SAMPLE_STORAGE_PROCESSING(EnumMixin, Enum):
+    ['storage_processing'],
+    ['term', 'label'],
+    ['organism_part'],
+    ['developmental_stage'],
```
Please, remember that changing value derived from a dictionary table should change the relation with that table, not the term itself, as I pointed out in `common.tasks.BatchUpdateMixin` comment

Alexey-ebi replied:
Removed possibility to change ontology terms, just show message to user. This should be changed in another process using Zooma.

bunop commented on django-data/image/common/tasks.py:
```diff
@@ -13,10 +13,53 @@
+        for id, value in ids.items():
+            if value == '' or value == 'None':
+                value = None
+            if item_type == 'sample':
```
In this common mixin you are referring to a lot of objects and using them in the same way. I suggest you to model this mixin without referring to UID models. For example:
```python
class BatchUpdateMixin:
    """Mixin to do batch update of fields to fix validation"""
    
    # item type will be a UID class that wil be set by child classes
    item_type = None
    
    # you could do the same for Submission class, in order to not have
    # image_model.classes inside your common.tasks

    def batch_update(self, submission_id, ids, attribute):
        # id is a reserved keyword. when you need id, the community suggest
        # to add an _ after the name
        for id_, value in ids.items():
            if value == '' or value == 'None':
                value = None
                
            # get your item object
            item_object = self.item_type.objects.get(pk=id_)
            
            if getattr(item_object, attribute) != value:
                setattr(item_object, attribute, value)
                item_object.save()

        # Update submission
        submission_obj = Submission.objects.get(pk=submission_id)
        submission_obj.status = NEED_REVISION
        submission_obj.message = "Data updated, try to rerun validation"
        submission_obj.save()

        send_message(
            submission_obj, construct_validation_message(submission_obj)
        )
```
 Then in `animal.tasks.BatchUpdateAnimals` you will declare the UID model as the class attribute:
```python
class BatchUpdateAnimals(MyTask, BatchUpdateMixin):
    name = "Batch update animals"
    description = """Batch update of field in animals"""
    
    # item type will be a UID class that wil be set by child classes
    # check that item_type is not a MyTask(celery.task( attribute 
    # (I suppose not)
    item_type = Animal

    # Stuff ...
    
    # you don't need to specify item_type parameter, since now is a class attribute
    def run(self, submission_id, animal_ids, attribute):
        """Function for batch update attribute in animals
        Args:
            submission_id (int): id of submission
            animal_ids (dict): dict with id and values to update
            attribute (str): attribute to update
        """

        # Stuff ...
```

bunop replied:
Sorry for this messy. I should have defined a UML design of this project, but unfortunately I have no time to learn UML, write documentations and think better on all functionalities of InjectTool

Alexey-ebi replied:
Done

bunop commented on django-data/image/submissions/views.py:
```diff
@@ -160,16 +173,117 @@ class SubmissionValidationSummaryView(OwnerMixin, DetailView):
+            editable = list()
+            for message in validation_summary.messages:
+                message = ast.literal_eval(message)
+                if uid2biosample(message['offending_column']) in \
```
Should we refactor `validation.models` to get a more easier description of what column as problem? consider inspecting this field for an unexpected issue in InjectTool using SQL

bunop commented on django-data/image/submissions/views.py:
```diff
@@ -160,16 +173,117 @@ class SubmissionValidationSummaryView(OwnerMixin, DetailView):
+                                        ])
+        self.offending_column = uid2biosample(
+            self.message['offending_column'])
+        if is_target_in_message(self.message['message'],
```
There are a lot of condition: can we simplify this stuff? will be difficult to maintain it if UID fields change (for instance, if `IMAGE-metadata` changes)

bunop replied:
In common constants, you have grouped related fields accordingly their information (birth location, collection place, ...). Maybe if you could code this with a django `forms.ModelForm` you could do additional checks and displaying fields easily. You could also validate forms before sending data to tasks. However this is my opinion, I don't know if it could be easly implemented in your batch update idea

Alexey-ebi replied:
Done

bunop commented on django-data/image/submissions/views.py:
```diff
@@ -160,16 +173,117 @@ class SubmissionValidationSummaryView(OwnerMixin, DetailView):
+            self.show_units = False
+            self.units = None
+            self.offending_column = 'term'
+            if self.offending_column == 'organism_part':
```
Here you refer to ontology relations. Please review this part accordingly the comment I made for `common.tasks.BatchUpdateMixin`

Alexey-ebi replied:
Removed possibility to change ontology terms, just show message to user. This should be changed in another process using Zooma.

bunop commented on django-data/image/submissions/views.py:
```diff
@@ -160,16 +173,117 @@ class SubmissionValidationSummaryView(OwnerMixin, DetailView):
+            elif self.offending_column == 'birth_location_accuracy':
+                self.units = [unit.name for unit in ACCURACIES]
+        elif is_target_in_message(self.message['message'],
+                                  ONTOLOGY_VALIDATION_MESSAGES):
```
Here you impose conditions on ontology messages. However, in such case you can only change an ontology with another one (consider `common.tasks.BatchUpdateMixin` comments). It is useful to hard code the ontology message patterns in `common.constants` and parse them with `submission.helpers` method? What happens if we need to update `IMAGE-ValidationTools` since it changes its error messages as we saw during development? In my opinion, the message you store in database should be enough to help user doing a supervised update (that will be changing term, or call zooma for instance). The term itself or the manual update of ontology ID should be managed by trusted user or in a different process as I stated in `common.tasks.BatchUpdateMixin` comment

Alexey-ebi replied:
Removed possibility to change ontology terms, just show message to user. This should be changed in another process using Zooma.

bunop commented on django-data/image/common/constants.py:
```diff
@@ -173,3 +173,34 @@ class SAMPLE_STORAGE_PROCESSING(EnumMixin, Enum):
+    ['species'],
+]
+
+ONTOLOGY_VALIDATION_MESSAGES = [
```
I added also a comment on `submission,views` regarding those message pattern

Alexey-ebi replied:
Removed possibility to change ontology terms, just show message to user. This should be changed in another process using Zooma.

bunop reviewed (commented):
Many thanks @Alexey-ebi for your contribution! I added some points to your code, maybe you could simplify something and have less code to maintain. You should write also some test code for your stuff. I understand that is not useful have 100% coverage, however trying to test your code or at least the essential parts, will help me to maintain the whole project since I could see where things stops working after changing other parts of this project.

Alexey-ebi committed:
Removing possibility to edit ontology terms and code refactoring

bunop committed:
:twisted_rightwards_arrows: Merge branch 'master' into validation_batch_update

Conflicts:
	django-data/image/animals/tasks.py
	django-data/image/samples/tasks.py
	django-data/image/submissions/urls.py
	django-data/image/submissions/views.py
	django-data/image/validation/models.py

all conflict are resolved

bunop committed:
:white_check_mark: test sample ValidationSummary changes after deleting animals (in batch)

bunop committed:
:recycle: refactor ValidationSummary and test construct_validation_message

bunop committed:
:bug: deal with missing offending_column message (for old validations)

bunop committed:
:white_check_mark: test SubmissionValidationSummaryFixErrorsView

bunop committed:
:white_check_mark: raise coverage in helpers method

bunop committed:
:white_check_mark: test sample/animal update tasks

bunop committed:
:recycle: refactor submissions tests

bunop committed:
:white_check_mark: test FixValidation view

bunop committed:
:recycle: remove target="_blank" for submission edit

bunop committed:
:white_check_mark: test SubmissionValidationSummaryFixErrorsView
